certbot/compat.py

```diff
@@ -0,0 +1,108 @@
+"""
+Compatibility layer to run certbot both on Linux and Windows.
+
+The approach used here is similar to Modernizr for Web browsers.
+We do not check the plateform type to determine if a particular logic is supported.
+Instead, we apply a logic, and then fallback to another logic if first logic
+is not supported at runtime.
+
+Then logic chains are abstracted into single functions to be exposed to certbot.
+"""
+import os
+import select
+import sys
+import errno
+import ctypes
+
+from certbot import errors
+
+try:
+    # Linux specific
+    import fcntl # pylint: disable=import-error
+except ImportError:
+    # Windows specific
+    import msvcrt # pylint: disable=import-error
+
+UNPRIVILEGED_SUBCOMMANDS_ALLOWED = [
+    'certificates', 'enhance', 'revoke', 'delete',
+    'register', 'unregister', 'config_changes', 'plugins']
+def raise_for_non_administrative_windows_rights(subcommand):
+    """
+    On Windows, raise if current shell does not have the administrative rights.
+    Do nothing on Linux.
+    """
+    # Why not simply try ctypes.windll.shell32.IsUserAnAdmin() and catch AttributeError ?
+    # Because windll exists only on a Windows runtime, and static code analysis engines
+    # do not like at all non existent objects when run from Linux (even if we handle properly
+    # all the cases in the code).
+    # So we access windll only by reflection to trick theses engines.
+    if hasattr(ctypes, 'windll') and subcommand not in UNPRIVILEGED_SUBCOMMANDS_ALLOWED:
+        windll = getattr(ctypes, 'windll')
+        if windll.shell32.IsUserAnAdmin() == 0:
+            raise ValueError(
```

Comment:
Can we use `errors.Error` here? We should have this better documented (and I created https://github.com/certbot/certbot/issues/6357 for us to do so), but in general, we try to always raise custom error types in Certbot if we expect them to crash the program.

Reply:
Done.